packages/ui/src/data/sqon/utils.ts

```diff
@@ -0,0 +1,200 @@
+};
+
+/**
+ * Convert a synthetic sqon into an executable sqon. Resolve all references if needed.
+ *
+ * @param {Array<ISyntheticSqon>} sqonsList All synthetic sqons in the query builder
+ * @param {ISyntheticSqon} syntheticSqon The synthetic sqon to resolve
+ *
+ * @returns {ISqonGroupFilter} An executable sqon
+ */
+export const resolveSyntheticSqon = (
+    sqonsList: ISyntheticSqon[],
+    syntheticSqon: ISyntheticSqon | Record<string, never>,
+): ISqonGroupFilter => {
+    const getNewContent = (syntheticSqon: ISyntheticSqon | Record<string, never>): TSqonContent => {
+        return syntheticSqon.content
+            .map((c: any) => (isReference(c) ? sqonsList[c] : c))
+            .map((c: any) => resolveSyntheticSqon(sqonsList, c));
+    };
+
+    if (isBooleanOperator(syntheticSqon)) {
+        return {
+            op: syntheticSqon.op,
+            content: getNewContent(syntheticSqon),
+        };
+    }
+
+    return {
+        op: syntheticSqon.op,
+        content: syntheticSqon.content as TSqonContent,
+    };
+};
+
+/**
+ * Remove recursively a sqon using its index.
+ *
+ * @param {number} indexToRemove Index of the synthetic sqon to remove
+ * @param {Array<ISyntheticSqon>} sqonsList All synthetic sqons in the query builder
+ *
+ * @returns {Array<ISyntheticSqon>} The new synthetic sqons list
+ */
+export const removeSqonAtIndex = (indexToRemove: number, sqonsList: ISyntheticSqon[]): Array<ISyntheticSqon> => {
+    const getNewContent = (indexToRemove: number, content: TSyntheticSqonContent) => {
+        return content
+            .filter((content: any) => content !== indexToRemove)
+            .map((s: any) => (isReference(s) ? (s > indexToRemove ? s - 1 : s) : s));
+    };
+
+    return sqonsList
+        .filter((s, i) => i !== indexToRemove)
+        .map((sqon) => {
+            if (isEmptySqon(sqon)) {
+                return sqon;
+            }
+
+            return {
+                ...sqon,
+                content: getNewContent(indexToRemove, sqon.content),
+            };
+        }).filter((s) => isNotEmptySqon(s));
+};
+
+/**
+ * Recursively change the operator throughout a given synthetic sqon
+ *
+ * @param {TSqonGroupOp} operator The new operator
+ * @param {ISyntheticSqon} syntheticSqon The synthetic sqon to update
+ *
+ * @returns {ISyntheticSqon} The modified synthetic sqon
+ */
+export const changeCombineOperator = (operator: TSqonGroupOp, syntheticSqon: ISyntheticSqon): ISyntheticSqon => {
+    const changeSubContentOperator = (content: TSyntheticSqonContent) => {
+        return content.map((subContent: any) => {
+            if (isBooleanOperator(subContent)) {
+                return changeCombineOperator(operator, subContent);
+            }
+
+            return subContent;
+        });
+    };
+
+    return {
+        ...syntheticSqon,
+        op: operator,
+        content: changeSubContentOperator(syntheticSqon.content),
+    };
+};
```

Comment:
I'm not convinced that the extra steps here make it easier to understand

```javascript
export const changeCombineOperator = (operator: TSqonGroupOp, syntheticSqon: ISyntheticSqon): ISyntheticSqon => ({
  ...syntheticSqon,
  op: operator,
  content: syntheticSqon.content.map((subContent: any) => isBooleanOperator(subContent) ? 
    changeCombineOperator(operator, subContent) :
    subContent)
})
```

Reply:
is there a way to avoid any here?

Reply:
I tried, but typescript doesn't like the fact that subContent can be : `Array<ISyntheticSqon | IValueFilter | number>`